### DISCUSSION TITLE
File not found error when file exists.

### DISCUSSION THREAD

**Discussion by eliphatfs** [seaweedfs/seaweedfs#5432] [Q&A]
I'm running a ML training script upon seaweedfs (mounted with `weed mount` for training data. Occasionally, I get FileNotFoundError at `open`. When I `cat` the file afterwards, the file prints without any errors.

There are errors like `E0326 00:37:31.023989 weedfs_dir_lookup.go:34 dir Lookup xxx/xxx: rpc error: code = Unavailable desc = error reading from server: read tcp 10.8.149.4:57890->10.8.149.13:18888: read: connection timed out` in the log. However, this shall result in a EIO error instead of a ENOENT during inode lookup (https://github.com/seaweedfs/seaweedfs/blob/master/weed/mount/weedfs_dir_lookup.go#L19). After inode lookup, the local store should be able to locate the file key.

Looking through seaweedfs source code I did not find a code path to return ENOENT in `open`.

During lookup each directory for inode: https://github.com/seaweedfs/seaweedfs/blob/master/weed/mount/weedfs_dir_lookup.go L39, L47, L55; During open handle from inode: https://github.com/seaweedfs/seaweedfs/blob/master/weed/mount/weedfs_filehandle.go L10 in maybeLoadEntry; but the filer store should be able to locate the file if the inode lookup is successful?

I guess this issue may be related with transient network errors between meta/master/volume/mount clients, but I cannot reproduce with an aritificial setup.

Any assistance would be appreciated. Thank you.

**System Setup**
- `/usr/local/bin/weed server -volume=0 -filer -dir=/weedfs`
- `/usr/local/bin/weed volume -max=400 -dir=/weedfs -mserver=10.8.149.13:9333` on 8 machines different from the master
- These commands are done by systemctl services with setups exactly as the wiki page.
- OS version: Ubuntu 22.04
- output of `weed version`: version 8000GB 3.63 54d7748a4a54d94a31ce04d05db801faeff4f690 linux amd64
- if using filer, show the content of `filer.toml`: no `filer.toml`
- security:
```
[access]
ui = false

[grpc]
ca = "/etc/ariesdockerd/certs/Aries_SeaweedFS_CA.crt"

[grpc.volume]
cert = "/etc/ariesdockerd/certs/volume01.crt"
key  = "/etc/ariesdockerd/certs/volume01.key"

[grpc.master]
cert = "/etc/ariesdockerd/certs/master01.crt"
key  = "/etc/ariesdockerd/certs/master01.key"

[grpc.filer]
cert = "/etc/ariesdockerd/certs/filer01.crt"
key  = "/etc/ariesdockerd/certs/filer01.key"

[grpc.client]
cert = "/etc/ariesdockerd/certs/client01.crt"
key  = "/etc/ariesdockerd/certs/client01.key"
```

**Comment by chrislusf** [comment]
Please share more details on the ML scripts file access pattern, throughput, file size, file count, read or write.

**Reply by eliphatfs** [reply]
~20M/s, ~200 files/s, read only, random access across a dataset organized as dataset-name/000(partition)/00000(object)/000.img(image)
For each batch (current bs=8) we read all 41 images of the 8 objects. objects are sampled uniformly at random.

**Reply by chrislusf** [reply]
Before the random access, could you run `tree .` or something similar to list all directories? Basically it can cache all directory listing. Let me know whether it helps.

**Reply by eliphatfs** [reply]
`tree .` itself fails halfway with I/O error...

I added external loop retries in lookup and read operations (first commits of `https://github.com/eliphatfs/seaweedfs/commits/aries/`).
The problem seems fixed by these retries.

I guess there is some edge case following a meta cache EIO that results in a spurious ENOENT.

The internal retry logic only retries when the string `transport` is found in errors, but for some reason transient network errors on my side do not lead to `transport` errors.

```
Mar 27 22:56:12 aries-00 seaweedfs-client[2245758]: E0327 22:56:12.563953 weedfs_file_sync.go:102 /xxx/yyy/zzz/www.json doFlush: flush data: upload data: filerGrpcAddress assign volume: rpc error: code = Unavailable desc = error reading from server: read tcp 10.8.149.4:42604->10.8.149.13:18888: read: connection timed out
Mar 27 22:57:54 aries-00 seaweedfs-client[2245758]: E0327 22:57:54.508121 meta_cache_subscribe.go:76 follow metadata updates: subscribing filer meta change: rpc error: code = Unavailable desc = error reading from server: read tcp 10.8.149.4:47652->10.8.149.13:18888: read: connection reset by peer
```

**Reply by chrislusf** [reply]
To reproduce this, could you please have code or a script to create directories and empty files similar to your use case? I can run a "tree" command to see the errors.

**Reply by eliphatfs** [reply]
The data is here https://tri-ml-public.s3.amazonaws.com/datasets/views_release.tar.gz
The key is not to reproduce the directory structure, but the special kind of connection errors i think.

**Reply by eliphatfs** [reply]
I see `read: connection reset by peer` and `read: connection timed out` types of error not undergoing the retry process. Only `transport: Error while dialing: dial tcp 10.8.149.13:18888: connect: network is unreachable` is retried.

**Reply by Material-Scientist** [reply]
You may have network connection issues. Make sure that all nodes can reach each other.
You can test by opening the weed shell with `weed shell -master="xxx:9333"` and running `cluster.check`.
